server/controllers/events/events_controller.go

```diff
@@ -713,3 +724,185 @@ func (e *VCSEventsController) commentNotAllowlisted(baseRepo models.Repo, pullNu
+	w.Header().Set("Content-Type", "application/json")
+
+	request, ctx, code, err := e.apiParseAndValidate(w, r)
+
+	if err != nil {
+		e.apiReportError(w, code, err)
+		return
+	}
+
+	cmds, err := request.getCommands(ctx, e.ProjectCommandBuilder.BuildPlanCommands)
+	if err != nil {
+		e.apiReportError(w, http.StatusInternalServerError, err)
+		return
+	}
+
+	defer e.Locker.UnlockByPull(ctx.HeadRepo.FullName, 0) // nolint: errcheck
+	var projectResults []command.ProjectResult
+	for _, cmd := range cmds {
+		res := e.ProjectPlanCommandRunner.Plan(cmd)
+		projectResults = append(projectResults, res)
+	}
+	result := command.Result{ProjectResults: projectResults}
+
+	if result.HasErrors() {
+		code = http.StatusInternalServerError
+	}
+
+	// TODO: make a better response
+	response, _ := json.Marshal(result)
+	e.respond(w, logging.Debug, code, string(response))
+}
+
+func (e *VCSEventsController) APIApply(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	request, ctx, code, err := e.apiParseAndValidate(w, r)
+
+	if err != nil {
+		e.apiReportError(w, code, err)
+		return
+	}
+
+	// We must first make the plan for all projects
+	cmds, err := request.getCommands(ctx, e.ProjectCommandBuilder.BuildPlanCommands)
+	if err != nil {
+		e.apiReportError(w, http.StatusInternalServerError, err)
+		return
+	}
+
+	defer e.Locker.UnlockByPull(ctx.HeadRepo.FullName, 0) // nolint: errcheck
+	var projectResults []command.ProjectResult
+	for _, cmd := range cmds {
+		res := e.ProjectPlanCommandRunner.Plan(cmd)
+		projectResults = append(projectResults, res)
+	}
+	result := command.Result{ProjectResults: projectResults}
+
+	// We can now prepare and run the apply step
+	cmds, err = request.getCommands(ctx, e.ProjectCommandBuilder.BuildApplyCommands)
+	if err != nil {
+		e.apiReportError(w, http.StatusInternalServerError, err)
+		return
+	}
+	projectResults = nil
+	for _, cmd := range cmds {
+		res := e.ProjectApplyCommandRunner.Apply(cmd)
+		projectResults = append(projectResults, res)
+	}
+	result = command.Result{ProjectResults: projectResults}
+
+	if result.HasErrors() {
+		code = http.StatusInternalServerError
+	}
+
+	response, _ := json.Marshal(result)
+	e.respond(w, logging.Debug, code, string(response))
+}
+
+func (e *VCSEventsController) apiParseAndValidate(w http.ResponseWriter, r *http.Request) (*APIRequest, *command.Context, int, error) {
+	if len(e.APISecret) == 0 {
+		return nil, nil, http.StatusBadRequest, fmt.Errorf("Ignoring request since API is disabled")
+	}
+
+	// Validate the secret token
+	header := "X-Atlantis-Token"
+	secret := r.Header.Get(header)
+	if secret != string(e.APISecret) {
+		return nil, nil, http.StatusUnauthorized, fmt.Errorf("header %s did not match expected secret", header)
+	}
+
+	// Parse the JSON payload
+	bytes, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, nil, http.StatusBadRequest, fmt.Errorf("failed to read request")
```

Comment:
```suggestion
		return nil, nil, http.StatusBadRequest, fmt.Errorf("Failed to read request")
```

Reply:
Good point on the consistency. I didn't pay attention to it since I was focused on resolving the conflicts. Updated all the error messages to lowercase since it's what our [style guide](https://github.com/runatlantis/atlantis/blob/master/CONTRIBUTING.md#errors) prefers.